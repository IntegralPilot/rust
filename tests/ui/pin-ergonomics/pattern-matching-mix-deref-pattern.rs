//@ revisions: normal pin_ergonomics deref_patterns both
//@ edition:2024
#![cfg_attr(any(pin_ergonomics, both), feature(pin_ergonomics))]
#![cfg_attr(any(deref_patterns, both), feature(deref_patterns))]
#![allow(incomplete_features)]

// This test verifies that the `pin_ergonomics` feature works well
// together with the `deref_patterns` feature under the error:
// "mix of deref patterns and normal constructors".

use std::pin::Pin;

struct Foo<T, U> {
    x: T,
    y: U,
}

struct Bar<T, U>(T, U);

enum Baz<T, U> {
    Foo(T, U),
    Bar { x: T, y: U },
}

fn foo_mut<T: Unpin, U: Unpin>(foo: Pin<&mut Foo<T, U>>) {
    match foo {
        Foo { x, y } => {}
        //[normal]~^ ERROR mismatched types
        //[deref_patterns]~^^ ERROR mix of deref patterns and normal constructors
        //[pin_ergonomics]~^^^ ERROR mix of deref patterns and normal constructors
        //[both]~^^^^ ERROR mix of deref patterns and normal constructors
        Pin { .. } => {}
    }
    let _ = || match foo {
        Foo { x, y } => {}
        //[normal]~^ ERROR mismatched types
        //[deref_patterns]~^^ ERROR mix of deref patterns and normal constructors
        //[pin_ergonomics]~^^^ ERROR mix of deref patterns and normal constructors
        //[both]~^^^^ ERROR mix of deref patterns and normal constructors
        Pin { .. } => {}
    };
}

fn foo_const<T: Unpin, U: Unpin>(foo: Pin<&Foo<T, U>>) {
    match foo {
        Foo { x, y } => {}
        //[normal]~^ ERROR mismatched types
        //[deref_patterns]~^^ ERROR mix of deref patterns and normal constructors
        //[pin_ergonomics]~^^^ ERROR mix of deref patterns and normal constructors
        //[both]~^^^^ ERROR mix of deref patterns and normal constructors
        Pin { .. } => {}
    }
    let _ = || match foo {
        Foo { x, y } => {}
        //[normal]~^ ERROR mismatched types
        //[deref_patterns]~^^ ERROR mix of deref patterns and normal constructors
        //[pin_ergonomics]~^^^ ERROR mix of deref patterns and normal constructors
        //[both]~^^^^ ERROR mix of deref patterns and normal constructors
        Pin { .. } => {}
    };
}

fn bar_mut<T: Unpin, U: Unpin>(bar: Pin<&mut Bar<T, U>>) {
    match bar {
        Bar(x, y) => {}
        //[normal]~^ ERROR mismatched types
        //[deref_patterns]~^^ ERROR mix of deref patterns and normal constructors
        //[pin_ergonomics]~^^^ ERROR mix of deref patterns and normal constructors
        //[both]~^^^^ ERROR mix of deref patterns and normal constructors
        Pin { .. } => {}
    }
    let _ = || match bar {
        Bar(x, y) => {}
        //[normal]~^ ERROR mismatched types
        //[deref_patterns]~^^ ERROR mix of deref patterns and normal constructors
        //[pin_ergonomics]~^^^ ERROR mix of deref patterns and normal constructors
        //[both]~^^^^ ERROR mix of deref patterns and normal constructors
        Pin { .. } => {}
    };
}

fn bar_const<T: Unpin, U: Unpin>(bar: Pin<&Bar<T, U>>) {
    match bar {
        Bar(x, y) => {}
        //[normal]~^ ERROR mismatched types
        //[deref_patterns]~^^ ERROR mix of deref patterns and normal constructors
        //[pin_ergonomics]~^^^ ERROR mix of deref patterns and normal constructors
        //[both]~^^^^ ERROR mix of deref patterns and normal constructors
        Pin { .. } => {}
    }
    let _ = || match bar {
        Bar(x, y) => {}
        //[normal]~^ ERROR mismatched types
        //[deref_patterns]~^^ ERROR mix of deref patterns and normal constructors
        //[pin_ergonomics]~^^^ ERROR mix of deref patterns and normal constructors
        //[both]~^^^^ ERROR mix of deref patterns and normal constructors
        Pin { .. } => {}
    };
}

fn main() {}
