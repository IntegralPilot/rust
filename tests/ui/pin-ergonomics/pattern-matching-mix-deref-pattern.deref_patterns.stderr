error: mix of deref patterns and normal constructors
  --> $DIR/pattern-matching-mix-deref-pattern.rs:27:9
   |
LL |         Foo { x, y } => {}
   |         ^^^^^^^^^^^^ matches on the result of dereferencing `Pin<&mut Foo<T, U>>`
...
LL |         Pin { .. } => {}
   |         ^^^^^^^^^^ matches directly on `Pin<&mut Foo<T, U>>`

error: mix of deref patterns and normal constructors
  --> $DIR/pattern-matching-mix-deref-pattern.rs:35:9
   |
LL |         Foo { x, y } => {}
   |         ^^^^^^^^^^^^ matches on the result of dereferencing `Pin<&mut Foo<T, U>>`
...
LL |         Pin { .. } => {}
   |         ^^^^^^^^^^ matches directly on `Pin<&mut Foo<T, U>>`

error: mix of deref patterns and normal constructors
  --> $DIR/pattern-matching-mix-deref-pattern.rs:46:9
   |
LL |         Foo { x, y } => {}
   |         ^^^^^^^^^^^^ matches on the result of dereferencing `Pin<&Foo<T, U>>`
...
LL |         Pin { .. } => {}
   |         ^^^^^^^^^^ matches directly on `Pin<&Foo<T, U>>`

error: mix of deref patterns and normal constructors
  --> $DIR/pattern-matching-mix-deref-pattern.rs:54:9
   |
LL |         Foo { x, y } => {}
   |         ^^^^^^^^^^^^ matches on the result of dereferencing `Pin<&Foo<T, U>>`
...
LL |         Pin { .. } => {}
   |         ^^^^^^^^^^ matches directly on `Pin<&Foo<T, U>>`

error: mix of deref patterns and normal constructors
  --> $DIR/pattern-matching-mix-deref-pattern.rs:65:9
   |
LL |         Bar(x, y) => {}
   |         ^^^^^^^^^ matches on the result of dereferencing `Pin<&mut Bar<T, U>>`
...
LL |         Pin { .. } => {}
   |         ^^^^^^^^^^ matches directly on `Pin<&mut Bar<T, U>>`

error: mix of deref patterns and normal constructors
  --> $DIR/pattern-matching-mix-deref-pattern.rs:73:9
   |
LL |         Bar(x, y) => {}
   |         ^^^^^^^^^ matches on the result of dereferencing `Pin<&mut Bar<T, U>>`
...
LL |         Pin { .. } => {}
   |         ^^^^^^^^^^ matches directly on `Pin<&mut Bar<T, U>>`

error: mix of deref patterns and normal constructors
  --> $DIR/pattern-matching-mix-deref-pattern.rs:84:9
   |
LL |         Bar(x, y) => {}
   |         ^^^^^^^^^ matches on the result of dereferencing `Pin<&Bar<T, U>>`
...
LL |         Pin { .. } => {}
   |         ^^^^^^^^^^ matches directly on `Pin<&Bar<T, U>>`

error: mix of deref patterns and normal constructors
  --> $DIR/pattern-matching-mix-deref-pattern.rs:92:9
   |
LL |         Bar(x, y) => {}
   |         ^^^^^^^^^ matches on the result of dereferencing `Pin<&Bar<T, U>>`
...
LL |         Pin { .. } => {}
   |         ^^^^^^^^^^ matches directly on `Pin<&Bar<T, U>>`

error: aborting due to 8 previous errors

